casper/src/main/scala/coop/rchain/casper/BlockCreator.scala

```diff
@@ -0,0 +1,237 @@
+                          parents,
+                          deploys,
+                          justifications,
+                          maxBlockNumber,
+                          shardId,
+                          version,
+                          now
+                        )
+                      } else {
+                        CreateBlockStatus.noNewDeploys.pure[F]
+                      }
+      signedBlock <- unsignedBlock.mapF(
+                      signBlock(_, dag, publicKey, privateKey, sigAlgorithm, shardId)
+                    )
+    } yield signedBlock
+
+  /*
+   * This mechanism is a first effort to make life of a deploying party easier.
+   * Instead of expecting the user to guess the current block number we assume that
+   * if no value is given (default: -1) rchain should try to deploy
+   * with the current known max block number.
+   *
+   * TODO: Make more developer friendly by introducing Option instead of a magic number
+   */
+  private def updateDeployHistory[F[_]: Sync: Log: Time: BlockStore](
+      state: Cell[F, CasperState],
+      maxBlockNumber: Long
+  ): F[Unit] = {
+    def updateDeployValidAfterBlock(deployData: DeployData, max: Long): DeployData =
+      if (deployData.validAfterBlockNumber == -1)
+        deployData.withValidAfterBlockNumber(max)
+      else
+        deployData
+
+    def updateDeployHistory(state: CasperState, max: Long): CasperState =
+      state.copy(deployHistory = state.deployHistory.map(deployData => {
+        updateDeployValidAfterBlock(deployData, max)
+      }))
+
+    state.modify(state => updateDeployHistory(state, maxBlockNumber))
+  }
+
+  // TODO: Remove no longer valid deploys here instead of with lastFinalizedBlock call
+  private def extractDeploys[F[_]: Monad: Log: Time: BlockStore](
+      dag: BlockDagRepresentation[F],
+      parents: Seq[BlockMessage],
+      maxBlockNumber: Long,
+      expirationThreshold: Int
+  )(implicit state: Cell[F, CasperState]): F[Seq[DeployData]] =
+    for {
+      state               <- state.read
+      currentBlockNumber  = maxBlockNumber + 1
+      earliestBlockNumber = currentBlockNumber - expirationThreshold
+      deploys             = state.deployHistory
+      validDeploys = deploys.filter(
+        d => notFutureDeploy(currentBlockNumber, d) && notExpiredDeploy(earliestBlockNumber, d)
+      )
+      deploysInCurrentChain <- DagOperations
+                                .bfTraverseF[F, BlockMessage](parents.toList)(
+                                  b =>
+                                    ProtoUtil
+                                      .unsafeGetParentsAboveBlockNumber[F](b, earliestBlockNumber)
+                                )
+                                .map { b =>
+                                  ProtoUtil.deploys(b).flatMap(_.deploy)
+                                }
+                                .toList
+    } yield (validDeploys -- deploysInCurrentChain.flatten).toSeq
```

Comment:
Depending on the size of the two lists the operator `--` can be expensive. If they are big it would be a good idea to construct a set from `deploysInCurrentChain` and then do `validDeploys.filter(! theSet.contains(_))`